### PR TITLE
React - PT 135 - Native button (Android mostly)

### PR DESCRIPTION
**Pull request by eduardb**
**TL;DR** Expose native UI components to react-native

**Solution - Android** To expose a native UI button, a `ReactButtonManager` was added. This is responsible instantiating the exposed component (`ReactButton`), and set its properties. For automeasuring the button based on its content, a subclass of `LayoutShadowNode` is used, `ReactButtonShadowNode`. There we listen for properties that are relevant for the size, and measuring happens in the `measure()` method. The code was copied from `ReactTextShadowNode` (with comments as well :joy:) and slightly adapted to our use case.

**RN side** A class wrapping the native component was created, with the properties exposed by it. The click listener is actually assigned to a `TouchableWithoutFeedback` element that wraps the native button, just to simplify things.

**Screenshot**

Sorry for the ugly buttons, we are not designers 💃

**On iOS, the component doesn't resize itself based on the content, that is something that it will be addressed in a feature PR, as this one was getting pretty big already anyway**

Paired/worked in parallel with @gbasile

### CODE REVIEW

File: ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButton.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+    public ReactButton(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
+        super(context, attrs, defStyleAttr, defStyleRes);
+        init();
+    }
+
+    private void init() {
+        setAllCaps(false);
```
Comment (gbasile) [inline ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButton.java]: Android weirdness :)

File: ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButton.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public void setBackgroundImage(String backgroundImage) {
```
Comment (gbasile) [inline ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButton.java]: The part about assets that have to be in the platform project is interesting. we should write about it in the doc

File: ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonShadowNode.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+    @Override
+    public void measure(
+            CSSNode node,
+            float width,
+            CSSMeasureMode widthMode,
+            float height,
+            CSSMeasureMode heightMode,
+            MeasureOutput measureOutput) {
+        // Code snippet copied from ReactTextShadowNode, and slightly adapted to our use case
```
Comment (gbasile) [inline ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonShadowNode.java]: I'm going to use this method for my NCU 1st Module :)

File: ReactNative/ReactTwitter/android/app/src/main/res/drawable/buttonbackground2.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+  <item android:state_enabled="false"
+    android:drawable="@drawable/button_background_disabled2" />
```
Comment (gbasile) [inline ReactNative/ReactTwitter/android/app/src/main/res/drawable/buttonbackground2.xml]: We can delete this one button-background-normal1.png

File: ReactNative/ReactTwitter/core/views/debug/debug-native-component-scene.js [hunk 5]
```diff
@@ -5,9 +5,11 @@ import {
   Navigator,
   StyleSheet,
   ListView,
-  Text
+  Alert
```
Comment (gbasile) [inline ReactNative/ReactTwitter/core/views/debug/debug-native-component-scene.js]: You don't need this alert anymore

**Comment by gbasile** [comment]
LGTM 👍

File: ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonShadowNode.java (hunk 3 above)
Comment (eduardb) [inline ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonShadowNode.java]: I think you can find worse in the RN source code

File: ReactNative/ReactTwitter/android/app/src/main/res/drawable/buttonbackground2.xml (hunk 4 above)
Comment (eduardb) [inline ReactNative/ReactTwitter/android/app/src/main/res/drawable/buttonbackground2.xml]: What do you mean by that?

Comment (gbasile) [inline ReactNative/ReactTwitter/android/app/src/main/res/drawable/buttonbackground2.xml]: there is an extra asset in this PR that is not used: button-background-normal1.png
We can delete it

File: ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonManager.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import android.graphics.Color;
+import android.util.TypedValue;
+
+import com.facebook.react.uimanager.BaseViewManager;
+import com.facebook.react.uimanager.ThemedReactContext;
+import com.facebook.react.uimanager.annotations.ReactProp;
+
+public class ReactButtonManager extends BaseViewManager<ReactButton, ReactButtonShadowNode> {
+
+    public static final String REACT_CLASS = "RCTNovodaButton";
```
Comment (danybony) [inline ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonManager.java]: This can be private

File: ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonManager.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    public String getName() {
+        return REACT_CLASS;
+    }
+
+    @Override
+    protected ReactButton createViewInstance(ThemedReactContext reactContext) {
+        return new ReactButton(reactContext);
+    }
+
+    @ReactProp(name = "enabled")
```
Comment (danybony) [inline ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonManager.java]: 👍

File: ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonShadowNode.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+            padding.set(0, 0, 0, 0);
+            return;
+        }
+        drawable.getPadding(this.padding);
+        markUpdated();
+    }
+
+    private Drawable getDrawableByName(String name) {
+        int resId = getThemedContext().getResources().getIdentifier(name.toLowerCase(), "drawable", getThemedContext().getPackageName());
+        if (resId <= 0) {
```
Comment (danybony) [inline ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonShadowNode.java]: ❌  from `getIdentifier()` documentation:
`int The associated resource identifier. Returns 0 if no such resource was found. (0 is not a valid resource ID.)`
So, `-1` for example is a valid id. You should check for 0 only

File: ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonShadowNode.java (hunk 3 above)
Comment (danybony) [inline ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonShadowNode.java]: Urgh, this is horrible...

File: ReactNative/ReactTwitter/core/views/debug/debug-native-component-scene.js [hunk 9]
```diff
@@ -43,7 +45,54 @@ var DebugNativeComponentScene = React.createClass({
   },
 
   _renderRow (rowData: string) {
-    return <Text style={styles.text}>{rowData}</Text>
+    let enabled = this._enabledForRowData(rowData)
+    let textColor = this._colorForRowData(rowData)
+    let textSize = this._sizeForRowData(rowData)
+    let backgrounImage = this._backgroundImageForRowData(rowData)
```
Comment (danybony) [inline ReactNative/ReactTwitter/core/views/debug/debug-native-component-scene.js]: `backgroundImage`

File: ReactNative/ReactTwitter/core/views/debug/debug-native-component-scene.js [hunk 10]
```diff
@@ -43,7 +45,54 @@ var DebugNativeComponentScene = React.createClass({
   },
 
   _renderRow (rowData: string) {
-    return <Text style={styles.text}>{rowData}</Text>
+    let enabled = this._enabledForRowData(rowData)
+    let textColor = this._colorForRowData(rowData)
+    let textSize = this._sizeForRowData(rowData)
```
Comment (danybony) [inline ReactNative/ReactTwitter/core/views/debug/debug-native-component-scene.js]: Would it be possible to group these in a style, and having then `_styleForRowData()`?

File: ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonShadowNode.java (hunk 8 above)
Comment (eduardb) [inline ReactNative/ReactTwitter/android/app/src/main/java/com/reacttwitter/widgets/ReactButtonShadowNode.java]: That's a very good catch!

File: ReactNative/ReactTwitter/core/views/debug/debug-native-component-scene.js (hunk 10 above)
Comment (eduardb) [inline ReactNative/ReactTwitter/core/views/debug/debug-native-component-scene.js]: They are not quite styles, but props, but I will take a look into a way of grouping them.

**Comment by danybony** [comment]
💯  LGTM! Nice one
